### ISSUE TITLE
Python syntax highlighting fails without window reload

### ISSUE THREAD

**Issue by tfire** [getcursor/cursor#1205]
Getting python syntax highlighting to apply to newly written code is requiring a window reload. 

- Remote SSH 
- Cursor Pyright + ms-python Python
- Amazon Linux 2 (Red Hat) 

If I type anything new, eg, `print()` and save the file, the following is logged:

```
2024-01-27 21:26:27.146 [error] [ms-python.python] provider FAILED
2024-01-27 21:26:27.146 [error] Error: No semantic tokens provided/needed for file:///home/ec2-user/my/python/file.py
	at /home/ec2-user/.cursor-server/extensions/ms-python.python-2023.14.0-universal/out/client/extension.js:2:2308663
	at re (/home/ec2-user/.cursor-server/extensions/ms-python.python-2023.14.0-universal/out/client/extension.js:2:2308957)
	at /home/ec2-user/.cursor-server/extensions/ms-python.python-2023.14.0-universal/out/client/extension.js:2:2303741
	at Immediate.<anonymous> (/home/ec2-user/.cursor-server/extensions/ms-python.python-2023.14.0-universal/out/client/extension.js:2:2303761)
	at process.processImmediate (node:internal/timers:476:21)
```

<img width="285" alt="image" src="https://github.com/getcursor/cursor/assets/3181512/dfc092bc-8c0d-49ce-a0d5-986c5fb2fbd3">

See the syntax highlighting above not applied to the last two prints.

**Comment by tfire** [comment]
```
2024-01-27 22:42:27.921 [info] [Error - 10:42:27 PM] Request textDocument/semanticTokens/full/delta failed.

```

**Comment by Sanger2000** [comment]
With Cursor Pyright, you should be using anysphere.python (rather than ms-python.python). Can help you figure this out via email if you send details to aman@anysphere.co

**Comment by tfire** [comment]
Thanks. Where is anysphere.python supposed to come from? It's not mentioned in any docs, across google, or in the extensions search. Taking this over email won't help others lacking the same information.

**Comment by arkaydeus** [comment]
In addition isort and Black Formatter rely on ms-python (preventing it being uninstalled).

Is there a decent solution here where we can still sort imports and format, while using Cursor Python?

Is there any reason why this hasn't been documented and explained?

**Comment by tfire** [comment]
I'm guessing anysphere.python is an unallowed fork of ms.python

**Comment by lukaemon** [comment]
> Thanks. Where is anysphere.python supposed to come from? It's not mentioned in any docs, across google, or in the extensions search. Taking this over email won't help others lacking the same information.

Did you find `anysphere.python`?

**Comment by arkaydeus** [comment]
Is there a solution to this. Python still looks completely broken in Cursor. One of the most popular languages out there completely unusable in Cursor. Surely there must be a solution?

**Comment by arkaydeus** [comment]
To summarise, cursor + cursor pyright + python doesn't work as syntax highlighting fails and that's a core part of the language server. It's basically failing to pickup tokens. Something is broken which means you can't really use the Python language in Cursor at the moment.

**Comment by lukaemon** [comment]
Yesterday Aman helped me resolved this python syntax/linting problem. Right now ms-python + Pylance works for me.

**Comment by arkaydeus** [comment]
Great. But please could you explain how you fixed it as it still doesn't work for other people.

**Comment by navels** [comment]
I would also like to know the magical incantation required to get syntax highlighting working.

**Comment by lukaemon** [comment]
Nothing at the client side. They pushed something in their backend that solved my problems.

**Comment by navels** [comment]
> Nothing at the client side. They pushed something in their backend that solved my problems.

lol it was probably `if username == "lukaemon"` . . .

**Comment by rian-dolphin** [comment]
Any update on this?

**Comment by jkryanchou** [comment]
Any update to this ? I have met the same issue.

**Comment by rian-dolphin** [comment]
I seem to have got it working though I'm not quite sure what did it. I did the following:

Uninstalled cursor by doing the following (on mac):
1. Quit the app if it's running.
2. Go to the Applications folder and move the app to the Trash.
3. Clean up associated files:
   - Open Finder
   - Press Cmd + Shift + G
   - Enter the following paths one by one and delete any folders or files related to the app:
     ~/Library/Application Support/
     ~/Library/Caches/
     ~/Library/Preferences/
     ~/Library/Saved Application State/
4. Empty the Trash.

Reinstalled cursor and selected the option to continue from scratch without any of my vscode settings.

Then the highlighting was working, BUT when I changed to the Dark Modern theme it stopped working again. I then switched back to Anysphere Dark and back to Dark Modern again and it was working.

As I said, I'm not sure what did it but hopefully it can help someone else.

**Comment by arkaydeus** [comment]
I'm STILL getting this issue when using the Dark+ theme.

How to recreate:

Fresh install of cursor.
Open a Python project
Add new import (not highlighted)
Use new import (not highlighted)

<img width="305" alt="image" src="https://github.com/user-attachments/assets/849e835d-3a20-45ba-bff1-1d70b48f7995">

<img width="591" alt="image" src="https://github.com/user-attachments/assets/62c2d9c1-e00a-4ca7-aa1f-6fd0fb330cde">

Note that if you close the file and open it again, it then correctly highlights:

<img width="300" alt="image" src="https://github.com/user-attachments/assets/a633c4fa-5f6f-4a8d-bd87-9730b7581dbc">

This is incredibly frustrating and has lasted over 6 months now. PLEASE can someone provide advice on how they were able to solve this?

**Comment by arkaydeus** [comment]
> BUT when I changed to the Dark Modern theme it stopped working again.

Can you confirm if you got it working persistently? For me when I restart, it correctly highlights when the file is first open or the window is refreshed, but it doesn't carry on working when I type new code. It looks like the deltas aren't being passed to the language server.

![image](https://github.com/user-attachments/assets/28c789c1-2b04-486a-a158-6ec6f0825292)

**Comment by arkaydeus** [comment]
OK, I've made some progress on this.

Currently running MacOS. I installed a new profile (user) on my machine and strangely all worked fine. What I noticed though, is that different versions of the extensions are installed.

![image](https://github.com/user-attachments/assets/4981c4b5-b0dd-41fa-aed9-461fca0035f2)
![image](https://github.com/user-attachments/assets/9fdb0977-b74e-4b1f-977e-f9b7f6ebdf3f)

Compare the two. The darwin-arm64 versions work fine. The universal versions don't and keep running into issues. They're also older.

![image](https://github.com/user-attachments/assets/8b231660-1954-4a5b-adfe-b4621f7427af)

Following up and checking the two extensions.json files you can see the working version on the right and the broken version on the left.

On the right the targetPlatform is ‘darwin-arm64’ which is correct.
On the left it’s ‘universal’ and then the version is 2024.5.1 and not 2024.12.3

This is incredibly frustrating. Same machine. Same OS. Just different user profile.

Would be really interested to see how I can force a particular architecture for extensions.

**Comment by 0x-chad** [comment]
+1 on this issue

**Comment by ilbertt** [comment]
+1

Happening for me too, really annoying considering the price of the subscription and the fact that bug reduces a lot the productivity

**Comment by arkaydeus** [comment]
Have you guys tried migrating users? Really annoying that you have to do that, but it did allow me to install arm64 versions by default which didn't break.

**Comment by arkaydeus** [comment]
For anyone who finds this issue, has an M series Mac and isn't sure how to fix it short term, got to the  .cursor folder for your user and the extensions folder. IF you find you have python extensions that end in `-universal`, then your system is installing the wrong versions. You need -darwin-arm64 versions.

While I couldn't find a way to force architecture for extensions, creating a new user on the machine and then installing Cursor in that user profile did fix the problem. Even a complete clean of the old user and re-install DIDNT fix it.

So if you can stomach migrating to a new user profile, it's likely you will be able to fix this.

Would love to know if there's a way to force architecture.

**Comment by ilbertt** [comment]
I will never create a new profile on my computer just to fresh-install Cursor and make it work with the extensions. It's really annoying that the cursor team is not taking care of this issue, considering how much the product costs.

I even emailed them explaining my issue, without ever getting a reply. I have not subscribed to the yearly plan only because of this issue.
Now that the monthly renewal is getting closer, I will definitely consider **NOT renewing my subscription** and go back to VSCode + Codeium, which never gave me issues and never significantly reduced my productivity like this bug does.

I hope Cursor will start listening to their **paying** customers at some point.

**Comment by arkaydeus** [comment]
There's probably a way to force architecture for plugins. Does anyone know how to do this?

**Comment by simonryf** [comment]
I finally got it working, though this is by far not ideal. No idea what happens to auto update etc. Also I have not yet worked intensively with this. So if I run into further errors I will post again here. However syntax highlighting works right a way and not only after restarting python language server.

@arkaydeus Kudos to you! I couldn't do it without your info here. However switching to another user in macos didn't solve it for me. So I tried and tried and tried...

System Info

```bash
❯ sw_vers
ProductName:        macOS
ProductVersion:     15.0.1
BuildVersion:       24A348

~
❯ cursor -v
0.42.3
949de58bd3d85d530972cac2dffc4feb9eee1e40
arm64
```

1. Backup current extensions / start from scratch

```
# Backup all Extensions
mv ~/.cursor/extensions /tmp/
mkdir ~/.cursor/extensions
```

2. Download python for darwin-arm64 via Browser and drag the .vsxi in cursor extensions UI

(If you ask how I got to this link, I had a gazillion browser tabs open and from somewhere i got a similar link and adapted at according to arkaydeus screenshot above)

https://marketplace.visualstudio.com/_apis/public/gallery/publishers/ms-python/vsextensions/python/2024.12.3/vspackage?targetPlatform=darwin-arm64


3. Check 

```bash
cat ~/.cursor/extensions/extensions.json
```

I have now "targetPlatform": "darwin-arm64" for ms-python

4. Download pylance via Browser and also dragged it into cursor extensions UI

https://marketplace.visualstudio.com/_apis/public/gallery/publishers/ms-python/vsextensions/vscode-pylance/2024.8.1/vspackage

-> This automatically installed also Cursor Pyright

Now my extensions.json looks a bit weird, but it works for me.

```json
[
  {
    "identifier": {
      "id": "ms-python.debugpy",
      "uuid": "4bd5d2c9-9d65-401a-b0b2-7498d9f17615"
    },
    "version": "2024.6.0",
    "location": {
      "$mid": 1,
      "path": "/Users/sryf/.cursor/extensions/ms-python.debugpy-2024.6.0-darwin-arm64",
      "scheme": "file"
    },
    "relativeLocation": "ms-python.debugpy-2024.6.0-darwin-arm64",
    "metadata": {
      "installedTimestamp": 1729945269792,
      "pinned": true,
      "source": "gallery",
      "id": "4bd5d2c9-9d65-401a-b0b2-7498d9f17615",
      "publisherId": "998b010b-e2af-44a5-a6cd-0b5fd3b9b6f8",
      "publisherDisplayName": "ms-python",
      "targetPlatform": "darwin-arm64",
      "updated": false,
      "isPreReleaseVersion": false,
      "hasPreReleaseVersion": false
    }
  },
  {
    "identifier": {
      "id": "ms-python.python"
    },
    "version": "2024.12.3",
    "location": {
      "$mid": 1,
      "path": "/Users/sryf/.cursor/extensions/ms-python.python-2024.12.3",
      "scheme": "file"
    },
    "relativeLocation": "ms-python.python-2024.12.3",
    "metadata": {
      "installedTimestamp": 1729945269791,
      "source": "vsix",
      "id": "f1f59ae4-9318-4f3c-a9b5-81b2eaa5f8a5",
      "publisherDisplayName": "ms-python",
      "publisherId": "998b010b-e2af-44a5-a6cd-0b5fd3b9b6f8",
      "isPreReleaseVersion": false,
      "hasPreReleaseVersion": false
    }
  },
  {
    "identifier": {
      "id": "anysphere.pyright",
      "uuid": "d990d7f1-60b6-4234-91a7-d146a42bb230"
    },
    "version": "1.1.327",
    "location": {
      "$mid": 1,
      "path": "/Users/sryf/.cursor/extensions/anysphere.pyright-1.1.327-universal",
      "scheme": "file"
    },
    "relativeLocation": "anysphere.pyright-1.1.327-universal",
    "metadata": {
      "installedTimestamp": 1729945578486,
      "pinned": true,
      "source": "gallery",
      "id": "d990d7f1-60b6-4234-91a7-d146a42bb230",
      "publisherId": "963cf79c-e108-4418-8858-b88cc178c014",
      "publisherDisplayName": "Anysphere",
      "targetPlatform": "universal",
      "updated": false,
      "isPreReleaseVersion": false,
      "hasPreReleaseVersion": false
    }
  },
  {
    "identifier": {
      "id": "ms-python.vscode-pylance"
    },
    "version": "2024.8.1",
    "location": {
      "$mid": 1,
      "path": "/Users/sryf/.cursor/extensions/ms-python.vscode-pylance-2024.8.1",
      "scheme": "file"
    },
    "relativeLocation": "ms-python.vscode-pylance-2024.8.1",
    "metadata": {
      "installedTimestamp": 1729945578486,
      "pinned": true,
      "source": "vsix"
    }
  }
]
```

5. Install Color -, File Icon - & Product Icon Themes via Cursor Extensions UI

I like:

- One Dark Pro
- Material Icon Theme
- Material Product Icons

(But that's up to your preference :-)

**Comment by zaradana** [comment]
What worked for me wast uninstalling and installing "Cursor Pyright" extention from cursor. I am using Dark Modern theme, not sure if that makes a difference.

**Comment by arkaydeus** [comment]
Really hope these stick. My issue before I switched user was that after a while it would update and download the universal one over the top of arm64 version.

Save all your working ones in a backup folder so you can refresh that if it overwrites.

**Comment by joeblackwaslike** [comment]
Nothing works for me, the truth is sadly cursor does not support python and theres no indication from the devs whether it ever will.  Sadly, this makes cursor a non starter for me at this time, but I would love to try again in the future if they decide to start caring about python.

**Comment by ilbertt** [comment]
> Nothing works for me, the truth is sadly cursor does not support python and theres no indication from the devs whether it ever will. Sadly, this makes cursor a non starter for me at this time, but I would love to try again in the future if they decide to start caring about python.

Definitely agree with you. I'll also probably not renew the subscription the next month

**Comment by xdxindustries** [comment]
@simonryf  Thank you, can confirm this is the thing that finally fixed it for me.  Thank you very much!

**Comment by zjy110-openai** [comment]
@simonryf this worked for me as well, thank you so much!

**Comment by joeblackwaslike** [comment]
I was able to finally fix this and have documented the steps in the following gist https://gist.github.com/joeblackwaslike/752b26ce92e3699084e1ecfc790f74b2.

**Comment by anuin-cat** [comment]
You can install the extension： < Kylin Python(with jedi language server) > and < Pylint > to replace < python > and < pylance > to resolve this issue.

**Comment by blackbalancef** [comment]
@joeblackwaslike thanks, work for me

**Comment by anuin-cat** [comment]
@joeblackwaslike Thanks. I noticed your method on the forum earlier, but it didn't seem work for me. I'm win platform, but I believe the issue might be on my end (as I have quite a few plugins).

**Comment by AzzzGoodFish** [comment]
@joeblackwaslike thanks!!!!

**Comment by lukaemon** [comment]
@joeblackwaslike works like magic. you are even more handsome than movie joe black.

**Comment by simonryf** [comment]
@joeblackwaslike

I haven't tried your way yet... However what versions of ms-python.python and ms-python.vscode-pylance do you currently have installed?

With my approach it doesn't get updated.

I have manually installed (download and dragged into cursor plugins window) these right now.
AFAIK newest possible versions for cursor 0.44.11 (vscode 1.93.1)

[https://marketplace.visualstudio.com/_apis/public/gallery/publishers/ms-python/vsextensions/python/2024.14.1/vspackage?targetPlatform=darwin-arm64](https://marketplace.visualstudio.com/_apis/public/gallery/publishers/ms-python/vsextensions/python/2024.14.1/vspackage?targetPlatform=darwin-arm64)

[https://marketplace.visualstudio.com/_apis/public/gallery/publishers/ms-python/vsextensions/vscode-pylance/2024.12.100/vspackage ](https://marketplace.visualstudio.com/_apis/public/gallery/publishers/ms-python/vsextensions/vscode-pylance/2024.12.100/vspackage )

_(This solved [https://github.com/microsoft/pyright/discussions/8608](https://github.com/microsoft/pyright/discussions/8608) for me. Note: pylance uses pyright under the hood.)_

**Comment by albcunha** [comment]
I´m facing the same problems. to me, what seemed to fix the problem was jupyter notebook extension, that was need to specifically be restarted on the extensions menu

**Comment by joeblackwaslike** [comment]
> [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/1109300?s=64&amp;u=0009eeb9b3629b44699caac884e67ad555b08d32&amp;v=4">@joeblackwaslike](https://github.com/joeblackwaslike?rgh-link-date=2025-01-15T12%3A58%3A08.000Z)
> 
> I haven't tried your way yet... However what versions of ms-python.python and ms-python.vscode-pylance do you currently have installed?
> 
> With my approach it doesn't get updated.
> 
> I have manually installed (download and dragged into cursor plugins window) these right now. AFAIK newest possible versions for cursor 0.44.11 (vscode 1.93.1)
> 
> [marketplace.visualstudio.com/_apis/public/gallery/publishers/ms-python/vsextensions/python/2024.14.1/vspackage?targetPlatform=darwin-arm64](https://marketplace.visualstudio.com/_apis/public/gallery/publishers/ms-python/vsextensions/python/2024.14.1/vspackage?targetPlatform=darwin-arm64)
> 
> [marketplace.visualstudio.com/_apis/public/gallery/publishers/ms-python/vsextensions/vscode-pylance/2024.12.100/vspackage ](https://marketplace.visualstudio.com/_apis/public/gallery/publishers/ms-python/vsextensions/vscode-pylance/2024.12.100/vspackage)
> 
> _(This solved [microsoft/pyright#8608](https://github.com/microsoft/pyright/discussions/8608) for me. Note: pylance uses pyright under the hood.)_

@simonryf The way I fixed things was to point cursor to the same extension marketplace as vscode.  The one cursor points to by default is problematic in that they legally can't include pyright/pylance due to licensing structures that forbid the official microsoft python extension from being used in forks of vscode (which is what cursor is).

This is why python is broken in cursor by default and you should likely never expect them to do what I did or they will be in big trouble.  They cannot even tell you to do what I did.  For personal use, I don't really care what microsoft thinks, and I don't think I'm alone in that. :)

**Comment by simonryf** [comment]
> > [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/1109300?s=64&amp;u=0009eeb9b3629b44699caac884e67ad555b08d32&amp;v=4">@joeblackwaslike](https://github.com/joeblackwaslike?rgh-link-date=2025-01-15T12%3A58%3A08.000Z)
> > I haven't tried your way yet... However what versions of ms-python.python and ms-python.vscode-pylance do you currently have installed?
> > With my approach it doesn't get updated.
> > I have manually installed (download and dragged into cursor plugins window) these right now. AFAIK newest possible versions for cursor 0.44.11 (vscode 1.93.1)
> > [marketplace.visualstudio.com/_apis/public/gallery/publishers/ms-python/vsextensions/python/2024.14.1/vspackage?targetPlatform=darwin-arm64](https://marketplace.visualstudio.com/_apis/public/gallery/publishers/ms-python/vsextensions/python/2024.14.1/vspackage?targetPlatform=darwin-arm64)
> > [marketplace.visualstudio.com/_apis/public/gallery/publishers/ms-python/vsextensions/vscode-pylance/2024.12.100/vspackage ](https://marketplace.visualstudio.com/_apis/public/gallery/publishers/ms-python/vsextensions/vscode-pylance/2024.12.100/vspackage)
> > _(This solved [microsoft/pyright#8608](https://github.com/microsoft/pyright/discussions/8608) for me. Note: pylance uses pyright under the hood.)_
> 
> [@simonryf](https://github.com/simonryf) The way I fixed things was to point cursor to the same extension marketplace as vscode. The one cursor points to by default is problematic in that they legally can't include pyright/pylance due to licensing structures that forbid the official microsoft python extension from being used in forks of vscode (which is what cursor is).
> 
> This is why python is broken in cursor by default and you should likely never expect them to do what I did or they will be in big trouble. They cannot even tell you to do what I did. For personal use, I don't really care what microsoft thinks, and I don't think I'm alone in that. :)

@joeblackwaslike Thanks for your explanation. Lately I have upgraded cursor and run into exactly this problem that (don't recall exactly) pylance complained it won't run on "unlicensed" fork.

I don't recall where but someone suggested to use BasedPyright [https://github.com/DetachHead/basedpyright](https://github.com/DetachHead/basedpyright). I'm using it now for a couple of weeks and must say that I'm very happy with it.

so no more Pylance(ms-python) and "only" Python(ms-python), Cursor Pyright(Anysphere) and BasedPyright(detachhead)

(Not sure why I have still enabled Cursor Pyright, maybe it's not even active?)

**Comment by yitianlian** [comment]
Same issue

**Comment by leerob** [comment]
Thanks for opening this issue! We've moved our community support to the [Cursor Forum](https://forum.cursor.com) where you can get help from both the team and other users.

If this is a specific bug that you can still reproduce, please open a new issue with clear reproduction steps on the forum.

Thanks you!
